tests/test_users.py

```diff
@@ -0,0 +1,201 @@
+        name = "test_sete_policy@test.com"
+        user = User(self.directory, name)
+
+        with self.subTest("A user's status is enabled when provisioned."):
+            self.assertEqual(user.status, 'Enabled')
+        with self.subTest("A user's status is disabled when user.disable is called."):
+            user.disable()
+            self.assertEqual(user.status, 'Disabled')
+        with self.subTest("A user's status is enabled when user.enabled is called."):
+            user.enable()
+            self.assertEqual(user.status, 'Enabled')
+
+    def test_roles(self):
+        name = "test_sete_policy@test.com"
+        test_roles = [(f"group_{i}", f"Policy_{i}") for i in range(5)]
+        roles = [Role.create(self.directory, *i).name for i in test_roles]
+        role_names, role_statements = zip(*test_roles)
+        role_names = sorted(role_names)
+        role_statements = sorted(role_statements)
+
+        user = User(self.directory, name)
+        with self.subTest("a user has the default_user roles when created."):
+            self.assertEqual(user.roles, ['default_user'])
+
+        user.remove_roles(['default_user'])
+        role_name, role_statement = test_roles[0]
+        with self.subTest("A user has one role when a role is added."):
+            user.add_roles([role_name])
+            self.assertEqual(user.roles, [role_name])
+
+        with self.subTest("An error is raised when adding a role a user already has."):
+            with self.assertRaises(ad.exceptions.BatchWriteException) as ex:
+               user.add_roles([role_name])
+               self.assertTrue(ex.response['Error']['Message'].endswith(
+                   "LinkNameAlreadyInUseException: link name "
+                   "R->FakeRole_0->test_sete_policy%40test.com is already in use"))
+
+        with self.subTest("An error is raised when adding a role that does not exist."):
+            with self.assertRaises(ad.exceptions.BatchWriteException) as ex:
+               user.add_roles(["ghost_role"])
+               self.assertTrue(ex.response['Error']['Message'].endswith("/ Roles / ghost_role\\' does not exist.'"))
+
+        with self.subTest("A user inherits a roles policies when a role is added to a user."):
+            self.assertListEqual(sorted(user.lookup_policies()), sorted([user.statement, role_statement]))
+
+        with self.subTest("A role is removed from user when remove role is called."):
+            user.remove_roles([role_name])
+            self.assertEqual(user.roles, [])
+
+        with self.subTest("A user has multiple roles when multiple roles are added to user."):
+            user.add_roles(role_names)
+            self.assertEqual(user.roles, role_names)
+
+        with self.subTest("A user inherits multiple role policies when the user has multiple roles."):
+            self.assertListEqual(sorted(user.lookup_policies()),
+                                 sorted([user.statement] + role_statements))
```

Comment:
I'd expect this to be only `role_statements` after all roles were removed and only `user.add_roles(role_names)` executed.  How are `role_statements` and `role_names` linked?

Reply:
role name is the name of the role, role statement is the contents of the policy attached to the role.